### PR TITLE
[Specification] Allow specifying a custom module map file

### PR DESCRIPTION
**Pull request by segiddins**
\c @kylef @mrackwitz 

- [x] ~~Lint to prevent module_map defined on a subspec.~~ depends on #233.
- [x] Changelog

### CODE REVIEW

**Comment by orta** [comment]
So this is a breaking change that would require a Specs repo update if someone pushes a pod from a beta/master to the repo, are there any other DSL attributes we should add if we're doing this?

**Comment by segiddins** [comment]
Didn't we already add the module_name attribute?

-Samuel E. Giddins

On Feb 14, 2015, at 9:42 AM, Orta <notifications@github.com> wrote:

So this is a breaking change that would require a Specs repo update if someone pushes a pod from a beta/master to the repo, are there any other DSL attributes we should add if we're doing this?

**Comment by orta** [comment]
Legit point, and it's being used, ignore me.

screen shot 2015-02-14 at 12 47 48 pm

**Comment by segiddins** [comment]
I think this was one of the huge benefits of switching to JSON

**Comment by mrackwitz** [comment]
So with subspecs, I see the following two cases:
* **Without custom modulemap**: The generated modulemap will inject the generated umbrella header, which takes care of importing all public headers, including those of subspecs. Importing the full Clang module, will give you everything. But you can also import implicitly derived submodules by using e.g. `MyPod.MySubspecHeader`.

* **With custom modulemap**: The custom modulemap will be used and injected by the build setting in the generated target. Because this is given, there is also no need to generate an umbrella header. All public headers, even those of the subspecs, need to be transitively imported or known to the modulemap by explicitly enumerating them. In consequence, the pod authors would need to make sure, that they provide a public header with precompiler-guards with conditions for header-availability for each subspec (and eventually have something like a sub-umbrella-header for each subspec), so that the resulting Clang module does include the fulll public interface, including the subspecs.

**Comment by segiddins** [comment]
@mrackwitz sure. I think specifying your own module map is like 'hard mode' -- you're responsible for it doing the things it should. I don't think we'll recommend that anyone use it -- it'll just be an option for those who _need_ it. As to dealing with subspecs -- sure, that'd be nice, but (a) that would all be inside CocoaPods/CocoaPods and (b) it's 1/2 unrelated to this -- it can be done separately.

**Comment by mrackwitz** [comment]
>but (a) that would all be inside CocoaPods/CocoaPods and (b) it's 1/2 unrelated to this -- it can be done separately.

We could i.e. go a whole different approach and allow customizing the generated modulemap in the DSL, so that you could add modulemap-declarations also on the subspec level.
It's a bit of *intentional* overthinking as it isn't based yet on actual issues. *Intentional* because changing the DSL back and forth is limited, so we should take care and allow an extensible design, which is compatible with existing features.
After all, this seems to be given, so from my side, we're definitively good to go for this approach. :+1:

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 ## Master
 
+##### Enhancements
+
+* Allow specifying a custom module map file.
```
Comment (segiddins) [inline CHANGELOG.md]: two spaces

**Comment by segiddins** [comment]
Other than the one CHANGELOG comment and the linting, this should be good to go once the work in CocoaPods is done and https://github.com/CocoaPods/Core/pull/233 has been finished.

**Comment by orta** [comment]
This'll need a rebase :+1:

**Comment by mrackwitz** [comment]
Done! With the change of 1375f0f, which affects how the linter is tested, I discovered a duplicated warning, introduced with #233, which I fixed right here on this branch.

File: lib/cocoapods-core/specification/linter.rb [hunk 2]
```diff
@@ -210,10 +210,6 @@ def _validate_authors(a)
               'from default')
           end
         end
-
-        if a.empty?
```
Comment (segiddins) [inline lib/cocoapods-core/specification/linter.rb]: why is this removed?

File: spec/specification/dsl_spec.rb [hunk 3]
```diff
@@ -253,6 +253,11 @@ module Pod
         @spec.module_name = 'Three20'
         @spec.attributes_hash['module_name'].should == 'Three20'
       end
+
+      it 'allows to specify a custom module map file' do
```
Comment (segiddins) [inline spec/specification/dsl_spec.rb]: `it 'allows specifying a custom module map file' do

File: lib/cocoapods-core/specification/linter.rb (hunk 2 above)
Comment (mrackwitz) [inline lib/cocoapods-core/specification/linter.rb]: 238a282: The `author` attribute is specified in the DSL as `:required => true`, so it's linted automatically and we can rely on that. :)

File: spec/specification/dsl_spec.rb (hunk 3 above)
Comment (mrackwitz) [inline spec/specification/dsl_spec.rb]: it's like all the others before and after that

**Comment by mrackwitz** [comment]
This should be then ready now. Anything else left?

File: spec/specification/dsl_spec.rb (hunk 3 above)
Comment (segiddins) [inline spec/specification/dsl_spec.rb]: well that's no reason to introduce more bad grammar :)

**Comment by segiddins** [comment]
Nope, this is good! :+1:

File: spec/specification/dsl_spec.rb (hunk 3 above)
Comment (mrackwitz) [inline spec/specification/dsl_spec.rb]: From a linguist point you may be right, but I'd prefer to keep everything consistent. So feel free to go for it and open an englishification-of-all-the-specs PR. :stuck_out_tongue_winking_eye:
